### ISSUE TITLE
docker_container cannot find installed docker-py

### ISSUE THREAD

**Issue by ghost** [ansible/ansible-modules-core#4246]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

**docker_container** module
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

No custom configuration.
##### OS / ENVIRONMENT

Using Ansible on OS X 10.11.5.

Docker version:

```
docker --version
Docker version 1.12.0-rc4, build e4a0dbc, experimental
```

docker-py version:

```
pip show docker-py

---
Metadata-Version: 2.0
Name: docker-py
Version: 1.8.1
Summary: Python client for Docker.
Home-page: https://github.com/docker/docker-py/
Author: UNKNOWN
Author-email: UNKNOWN
Installer: pip
License: UNKNOWN
Location: /Users/rick/tmp/ansible/module-tests/docker_container/env/lib/python2.7/site-packages
Requires: ipaddress, six, requests, backports.ssl-match-hostname, websocket-client
Classifiers:
  Development Status :: 4 - Beta
  Environment :: Other Environment
  Intended Audience :: Developers
  Operating System :: OS Independent
  Programming Language :: Python
  Programming Language :: Python :: 2.6
  Programming Language :: Python :: 2.7
  Programming Language :: Python :: 3.3
  Programming Language :: Python :: 3.4
  Topic :: Utilities
  License :: OSI Approved :: Apache Software License
```
##### SUMMARY

I want to build a custom container on top of a base OS container. So I am trying to start a base OS container.
##### STEPS TO REPRODUCE

I came across this issue in my normal work environment on my Mac. I reproduced it in a clean environment created with virtualenv.

Here is the playbook I reproduced it with.

```
- name: start an OS image
  hosts: all
  connection: local

  tasks:

  - name: pull the OS image
    docker_container:
      name: my-test-container
      image: ubuntu:16.04
      state: started
```
##### EXPECTED RESULTS

Docker container would be started.
##### ACTUAL RESULTS

Here is the verbose failure output of the playboook run.

```
fatal: [localhost]: FAILED! => 
{
  "changed": false, 
   "failed": true, 
   "invocation": {
     "module_args": {
       "api_version": null, 
       "blkio_weight": null, 
       "cacert_path": null, 
       "capabilities": null, 
       "cert_path": null, 
       "command": null, 
       "cpu_period": null, 
       "cpu_quota": null, 
       "cpu_shares": null, 
       "cpuset_cpus": null, 
       "cpuset_mems": null, 
       "debug": false, 
       "detach": true, 
       "devices": null, 
       "dns_opts": null, 
       "dns_search_domains": null, 
       "dns_servers": null, 
       "docker_host": null, 
       "entrypoint": null, 
       "env": null, 
       "etc_hosts": null, 
       "exposed_ports": null, 
       "filter_logger": false, 
       "force_kill": false, 
       "groups": null, 
       "hostname": null, 
       "image": "ubuntu:16.04", 
       "interactive": false, 
       "ipc_mode": null, 
       "keep_volumes": true, 
       "kernel_memory": null, 
       "key_path": null, 
       "kill_signal": null, 
       "labels": null, 
       "links": null, 
       "log_driver": "json-file", 
       "log_options": null, 
       "mac_address": null, 
       "memory": "0", 
       "memory_reservation": null, 
       "memory_swap": null, 
       "memory_swappiness": null, 
       "name": "my-test-container", 
       "network_mode": null, 
       "networks": null, 
       "oom_killer": null, 
       "paused": false, 
       "pid_mode": null, 
       "privileged": false, 
       "published_ports": null, 
       "pull": false, 
       "read_only": false, 
       "recreate": false, 
       "restart": false, 
       "restart_policy": null, 
       "restart_retries": 0, 
       "security_opts": null, 
       "shm_size": null, 
       "ssl_version": null, 
       "state": "started", 
       "stop_signal": null, 
       "stop_timeout": null, 
       "timeout": null, 
       "tls": null, 
       "tls_hostname": null, 
       "tls_verify": null, 
       "trust_image_content": false, 
       "tty": false, 
       "ulimits": null, 
       "user": null, 
       "uts": null, 
       "volume_driver": null, 
       "volumes": null, 
       "volumes_from": null
    }, 
    "module_name": "docker_container"
  }, 
  "msg": "Failed to import docker-py - No module named requests.exceptions. Try `pip install docker-py`"
}
```

I am able to start the same container using the **shell** modue.

```
- name: start an OS image
  hosts: all
  connection: local

  tasks:

  - name: pull the OS image
    shell: |
      docker run --rm ubuntu:16.04
```

Here is the success output from that playbook.

```
changed: [localhost] => 
{
  "changed": true, 
   "cmd": "docker run --rm ubuntu:16.04", 
   "delta": "0:00:00.625392", 
   "end": "2016-07-24 17:34:22.476927", 
   "invocation": {
     "module_args": {
        "_raw_params": "docker run --rm ubuntu:16.04",
         "_uses_shell": true, 
         "chdir": null, 
         "creates": null, 
         "executable": null, 
         "removes": null, 
         "warn": true
      },
      "module_name": "command"
     }, 
   "rc": 0, 
   "start": "2016-07-24 17:34:21.851535", 
   "stderr": "", 
   "stdout": "", 
   "stdout_lines": [], 
   "warnings": []
}
```

**Comment by chouseknecht** [comment]
@rickmendes I'm testing with the stable-2.1 branch, and everything is working fine. I tried on both Centos 7 and OSX. docker-py 1.8.0 is installed in both environments, and both are running Docker 1.11.2. No problems. I also upgraded docker-py to 1.8.1 on the Centos box, and still seems to be OK.

Can you try running from source using stable-2.1?

Here's some testing results...

```
ansible 2.1.1.0 (stable-2.1 02389d6c51) last updated 2016/07/23 20:36:52 (GMT -400)
  lib/ansible/modules/core: (detached HEAD e09ab5bda3) last updated 2016/07/23 20:37:06 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 24812da6e8) last updated 2016/07/23 20:37:06 (GMT -400)
  config file =
  configured module search path = Default w/o overrides
```

Playbook:

```
- name: start an OS image
  hosts: localhost
  connection: local
  gather_facts: no
  tasks:
  - name: pull the OS image
    docker_container:
      name: my-test-container
      image: python:2.7
      command: /bin/sleep 600
```

Run the playbook:

```


PLAY [start an OS image] *******************************************************

TASK [pull the OS image] *******************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```

Check that container is running:

```
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
556c156db0d6        python:2.7          "/bin/sleep 600"    7 minutes ago       Up 53 seconds                           my-test-container
```

**Comment by ghost** [comment]
I was looking at the docker_common.py file and noticed the first import it does is from requests.exceptions.

I have the requests package too so this is more befuddling.

```
pip show  requests
---
Metadata-Version: 2.0
Name: requests
Version: 2.10.0
Summary: Python HTTP for Humans.
Home-page: http://python-requests.org
Author: Kenneth Reitz
Author-email: me@kennethreitz.com
Installer: pip
License: Apache 2.0
Location: /Users/rick/tmp/ansible/module-tests/docker_container/env/lib/python2.7/site-packages
Requires:
Classifiers:
  Development Status :: 5 - Production/Stable
  Intended Audience :: Developers
  Natural Language :: English
  License :: OSI Approved :: Apache Software License
  Programming Language :: Python
  Programming Language :: Python :: 2.6
  Programming Language :: Python :: 2.7
  Programming Language :: Python :: 3
  Programming Language :: Python :: 3.3
  Programming Language :: Python :: 3.4
  Programming Language :: Python :: 3.5
  Programming Language :: Python :: Implementation :: CPython
  Programming Language :: Python :: Implementation :: PyPy
```

**Comment by ghost** [comment]
@chouseknecht 

I cloned the source.

```
source ./hacking/env-setup

ansible --version
ansible 2.1.1.0 (stable-2.1 02389d6c51) last updated 2016/07/24 22:55:31 (GMT -700)
  lib/ansible/modules/core: (detached HEAD b3c606047d) last updated 2016/07/24 22:49:00 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD d9b4613941) last updated 2016/07/24 22:49:05 (GMT -700)
  config file =
  configured module search path = Default w/o overrides
```

Modified my playbook to pull the same image.

```
- name: start an OS image
  hosts: all
  connection: docker

  tasks:

  - name: pull the OS image
    docker_container:
      name: my-test-container
      image: python:2.7
      state: started
      command: /bin/sleep 300
```

Tried running the playbook from the source dir.

```
export ANSIBLE_INVENTORY=~/tmp/ansible/module-tests/docker_container/hosts

ansible-playbook ~/tmp/ansible/module-tests/docker_container/test-playbook.yml

PLAY [start an OS image] *******************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [pull the OS image] *******************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to import docker-py - No module named requests.exceptions. Try `pip install docker-py`"}

NO MORE HOSTS LEFT *************************************************************
    to retry, use: --limit @/Users/rick/tmp/ansible/module-tests/docker_container/test-playbook.retry

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

The big difference I see between your test and mine is the docker version. I am on a project that is going out on docker 1.12, which is why I am testing with docker 1.12.0.rc4.

**Comment by ghost** [comment]
@chouseknecht 

One more more thing. I tested with connection both local and docker and got the same result.

**Comment by ghost** [comment]
Working on getting you more info. So trying vagrant boxes with debian/jessie.

**vagrant 1**

ansible version:

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```

docker version:

```
docker --version
Docker version 1.11.2, build b9f10c9
```

playbook:

```
---

- name: start an OS image
  hosts: all
  connection: local
  gather_facts: False
  become: True

  tasks:

  - name: pull the OS image
    docker_container:
      name: my-test-container
      image: python:2.7
      state: started
      command: ["sleep", "600"]
```

playbook output:

```
PLAY [start an OS image] *******************************************************

TASK [pull the OS image] *******************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```

running container:

```
sudo docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
5d9460c90397        python:2.7          "sleep 600"         8 seconds ago       Up 7 seconds                            my-test-container
```

So with 1.11.2 we are good.

I'll be back with vagrant 2 info.

**Comment by ghost** [comment]
**vagrant 2**

ansible version:

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```

docker version:

```
docker --version
Docker version 1.12.0-rc4, build e4a0dbc
```

playbook:

```
- name: start an OS image
  hosts: all
  connection: local
  gather_facts: False
  become: True

  tasks:

  - name: pull the OS image
    docker_container:
      name: my-test-container
      image: python:2.7
      state: started
      command: ["sleep", "600"]
```

playbook output:

```
PLAY [start an OS image] *******************************************************

TASK [pull the OS image] *******************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```

running container:

```
sudo docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
80e1bbeff942        python:2.7          "sleep 600"         19 seconds ago      Up 18 seconds                           my-test-container
```

So it has no problems with 1.12 on Linux. It appears the issue is OSX related only as I was able to reproduce the error on another Mac today.

I will happily move my development to Linux with or without Vagrant. Should we close this or keep it open for others with the issue on OSX?

**Comment by chouseknecht** [comment]
I think we can close it. I don't think there is anything to do on our end.

**Comment by waddles** [comment]
> ```
> "msg": "Failed to import docker-py - No module named requests.exceptions. Try `pip install docker-py`"
> }
> ```

I struck this error message too using OSX then realised the (remote) inventory_hostname in play did not have docker installed yet. 👊

**Comment by ghost** [comment]
@waddles That makes sense to me. However, I can still not make it work with 'hosts: localhost', which is my laptop with docker installed.

**Comment by chouseknecht** [comment]
@rickmendes error message? version of ansible and version of docker-py installed on *localhost*?

**Comment by ghost** [comment]
@chouseknecht 

```
ansible --version
ansible 2.1.1.0
  config file =
  configured module search path = Default w/o overrides
```

```
pip freeze
ansible==2.1.1.0
backports.ssl-match-hostname==3.5.0.1
boto3==1.3.1
botocore==1.4.39
cffi==1.7.0
cryptography==1.4
docker-py==1.9.0
docutils==0.12
enum34==1.1.6
futures==3.0.5
idna==2.1
ipaddress==1.0.16
Jinja2==2.8
jmespath==0.9.0
MarkupSafe==0.23
paramiko==2.0.2
pyasn1==0.1.9
pycparser==2.14
pycrypto==2.6.1
python-dateutil==2.5.3
PyYAML==3.11
requests==2.11.0
six==1.10.0
websocket-client==0.37.0
```

playbook run

```
PLAY [start an OS image] *******************************************************

TASK [pull the OS image] *******************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to import docker-py - No module named requests.exceptions. Try `pip install docker-py`"}

NO MORE HOSTS LEFT *************************************************************
 [WARNING]: Could not create retry file 'test-playbook.retry'.         [Errno 2] No such file or directory: ''


PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1
```

**Comment by ozbillwang** [comment]
got same error in:

```
vagrant box: ubuntu/trusty64
ansible 2.1.1.0
Docker version 1.12.0, build 8eab29e
```

Here is the error:

```
TASK [ldap : Start the OpenLdap container] ****************************
fatal: [192.19.8.12]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to import docker-py - cannot import name NotFound. Try `pip install docker-py`"}
```

In fact the package has been installed on the client: 

```
$ sudo pip list |grep docker
docker-py (1.1.0)
```

**Comment by chouseknecht** [comment]
@SydOps the min docker-py version is 1.7.0.  https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/docker_common.py#L47

**Comment by 0angelic0** [comment]
I also have the problem here on OS X 10.11.6. The remote host is already installed docker-engine.

```
$ pip list | grep docker
docker-compose (1.8.0)
docker-py (1.9.0)
dockerpty (0.4.1)
```

```
TASK [Pull docker image nginx] *************************************************
fatal: [cmc-c-game-nginx-1]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to import docker-py - No module named requests.exceptions. Try `pip install docker-py`"}
```

**Comment by jhnlsn** [comment]
I'm getting the same error as everyone else here:

```
Docker version 1.12.1, build 23cf638
Ansible: 2.1.1.0
docker-py (1.9.0)
```

```
fatal: [127.0.0.1]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to import docker-py - No module named requests.exceptions. Try `pip install docker-py`"}
```

Downgrading back to `Ansible: 2.1.0.0` does not throw the error.

**Comment by deddu** [comment]
same error here

```
Docker version 1.12.0, build 8eab29e
ansible 2.1.1.0
docker-compose (1.8.0)
docker-py (1.9.0)
dockerpty (0.4.1)
```

Tried removing brew ansible and using pip installed one, no diffs.

**Comment by adityamarella** [comment]
I solved this by explicitly setting the PYTHONPATH to the following on Mac OSX
`export PYTHONPATH=/usr/local/lib/python2.7/site-packages/:$PYTHONPATH`

Most likely ansible/hacking/env-setup is not setting the python path correctly.

**Comment by waddles** [comment]
@adityamarella - just confirming I also have that set, so we now have 2 solutions:
1. ensure docker is installed on remote host
2. ensure PYTHONPATH is set and docker-py exists in that path

**Comment by DickChesterwood** [comment]
Same error here, but on Linux (Amazon AMI) - PYTHONPATH seems to have no effect.

Docker version 1.11.2, build b9f10c9/1.11.2
ansible 2.1.1.0
docker-compose (1.8.1)
docker-py (1.10.3)
docker-pycreds (0.2.1)
dockerpty (0.4.1)

I don't know python so I'm struggling badly but it seems to be some version mismatch of python interpreters...

**Comment by davegallant** [comment]
I am having the same issue on Ubuntu 14.04 LTS and a virtualenv:
ansible==2.1.1.0
docker-py==1.10.3

**Comment by DickChesterwood** [comment]
(Amazon AMI) I've been fighting with various combinations of pip (installed in various different ways), not realising I've been fighting in the wrong place - this has to be done on the TARGET node, not on the Ansible control code (it says "docker-py required on the node where the module is run", but when you're new to Ansible and don't know about the internal workings, it's a bit.... vague). Dim of me nonetheless.

So, on my target node, this is the recipe I've cooked up...

sudo easy_install pip
sudo /usr/local/bin/pip install docker-py

Check with python -c "import docker"

Playbook version:

```
 - name: Check Docker
   yum: name=docker state=latest
   become: true

 - name: Install pip
   easy_install: name=pip state=latest
   become: true

 - name: Install docker-py
   pip: name=docker-py executable=/usr/local/bin/pip
   become: true
```

(yum installing pip was just not working, the pip from that seems to be installing for python 2.6. Don't understand that).

Which has moved me on to "Error: docker-py version is 1.10.3. Minimum version required is 1.7.0." - go figure. I see there's a thread for that....https://github.com/ansible/ansible/issues/17495

(edit to add, due to the issue above I've found it necessary to downgrade to 1.9 of docker py, so pip install 'docker-py==1.9.0' --force-reinstall, or pip: name=docker-py executable=/usr/local/bin/pip version=1.9.0 on the playbook)

**Comment by davegallant** [comment]
Oh, you're right @DickChesterwood. Ansible Docker requires docker-py "on host that executes module".  This whole time, I thought that meant the Ansible host. I really need to check what is underneath the hood of Ansible.

**Comment by DickChesterwood** [comment]
Yeah - of course the module is executing on the remote target when you think about, but it's not that obvious from the outset! I feel a bit silly but at least it's working now.

**Comment by Freyert** [comment]
I thought the behavior of Ansible was that it _packaged_ all the dependencies necessary to run a module on a remote host and then shipped it up. Thus negating the need to install and manage module versions across all the hosts.

---

but yeah, if you do install docker-py on the servers you're managing it changes the error message to a crazy one.

**Comment by josephgimenez** [comment]
I ran into this issue as well, regarding the Docker module not importing.  In my case, it was because Ansible was calling a different python interpreter on my machine than I was expecting.  Ansible seems to default to using /usr/bin/python (system default python interpreter), whereas I was expecting it to use the interpreter installed by homebrew at /usr/local/bin/python.

The workaround I used was to update my inventory file and specify:
`localhost ansible_connection=local ansible_python_interpreter=/usr/local/bin/python`

This caused Ansible to begin using the correct intrepreter (which actually has the Docker module installed), and all was well.

**Comment by ozbillwang** [comment]
Thanks to point out the root cause, @asyncsrc 

http://docs.ansible.com/ansible/intro_inventory.html
https://github.com/ansible/ansible/issues/6345

```
ansible_python_interpreter

The target host python path. This is useful for systems with more than one Python or not located at /usr/bin/python such as *BSD, or where /usr/bin/python is not a 2.X series Python. We do not use the /usr/bin/env mechanism as that requires the remote user’s path to be set right and also assumes the python executable is named python, where the executable might be named something like python2.6.
```

**Comment by matiaslespiau** [comment]
I'm having the same problem in Mac OSX El Capitan 10.11.6
docker-py (1.10.6)
ansible 2.2.0.0

**Comment by pshahid** [comment]
Hi I experienced this issue on a Fedora 22 VM today. For me the fix was to upgrade docker-py on the Fedora VM via pip install. This may not be a solution for everyone, and I can confirm that docker-py did exist beforehand. Seems like it could be an issue between docker-py 1.10.0 and 1.10.6?

This is on the devel branch of ansible btw.

```
# Earlier on...
"msg": "Failed to import docker-py - cannot import name NotFound. Try `pip install docker-py`"

# But it exists in the VM...
$ pip list | grep docker
docker-compose (1.2.0)
docker-py (1.1.0)
docker-registry-core (2.0.3)
dockerpty (0.3.3)

$ sudo pip install -U docker-py
...spam...
Successfully installed docker-py-1.10.6 docker-pycreds-0.2.1 ipaddress-1.0.17 requests-2.12.4 six-1.10.0 websocket-client-0.40.0

# No longer an error
```

Hope this helps

**Comment by darkn3rd** [comment]
Not able to verify fix in Ansible 2.2 as it breaks for another reason on Mac OS X:

```
Traceback (most recent call last):\n  File \"/var/folders/79/4dttv20s6p30d78_yqmr3jjh0000gn/T/ansible_c_jdw536/ansible_module_docker_container.py\", line 652, in <module>\n    from ansible.module_utils.docker_common import *\n  File \"/var/folders/79/4dttv20s6p30d78_yqmr3jjh0000gn/T/ansible_c_jdw536/ansible_modlib.zip/ansible/module_utils/docker_common.py\", line 26, in <module>\nImportError: No module named 'urlparse'\n
```

**Comment by pshahid** [comment]
@darkn3rd are you using py3 by chance? urlparse moved to urllib.parse (see top of: https://docs.python.org/2/library/urlparse.html) as part of a py3 breaking change from py2. It seems as though whichever host you're running ansible on has python3 in the PATH ahead of py2 (or no py2 at all). You may want to try with python2 instead, as it seems ansible does not advertise full support for python3 yet: https://docs.ansible.com/ansible/python_3_support.html

Example of py2 vs py3 urlparse:

```
$ python --version
Python 3.5.2

$ python -c 'import urlparse'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named 'urlparse'

$ python --version
Python 2.7.12

$ python -c 'import urlparse'
```

**Comment by darkn3rd** [comment]
I am not trying to use Python3.  Ansible installed through `pip install` uses the System Libraries python 3.5, and ignores python installed in the path.

```
$ python --version
Python 2.7.12
$ python -c 'import urlparse'
$ echo $?
0
$ which python
/usr/local/bin/python
```

It seems after installing Ansible `sudo -H pip install ansible`, it is hard coded to use: `/usr/local/opt/python3/bin/python3.5`

Found out this may be homebrew issue: wrote up https://github.com/Homebrew/homebrew-python/issues/378

**Comment by pshahid** [comment]
@darkn3rd definitely an env issue, may be related to how you've setup homebrew. 

The following commands will tell you what the deal is:
```
$  pip --version
pip 8.1.2 from /usr/local/lib/python2.7/site-packages (python 2.7)
$ which pip
/usr/local/bin/pip
$ echo $PATH
/Users/<me>/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin
```

I'm guessing you used Python 3's pip to install ansible, instead of python 2's pip. One or the other is going to be getting picked from your $PATH. You may not even have python 2 installed to begin with, though, so there's that to consider.

**Comment by pshahid** [comment]
# TL;DR

Settings ansible_python_interpreter is a workaround

```
$ cat hosts
[test]
localhost ansible_python_interpreter=/usr/local/bin/python
```

But why is that needed?! Don't know enough about ansible runtime to understand yet.

## Detail

### Repro artifacts

``ansible-playbook -i hosts myplaybook.yaml -vvvv --flush-cache``

```
 cat myplaybook.yaml 
---
- hosts: all
  connection: local

  tasks:

  - name: pull the OS image
    docker_container:
      name: my-test-container
      image: ubuntu:16.04
      state: started
```

```
cat hosts
[test]
localhost
```

### Versions

OSX 10.11.4
python 2.7.12
docker-py 1.10.6
requests 2.12.2
ansible 2.3.0 (devel)
docker 1.12.5, build 7392c3b

### Description

Ok so I did a little bit of digging on this error and I found that ansible is, strangely, not using the environment that I call it with. I have no idea why it would think that's what the user (invoker) wants;  I could install it in any virtualenv, vm, docker image, and run ansible that way and it should respect the caller's environment.

Here's what I observed:

I downloaded ansible @ HEAD of devel ran ``python setup.py install`` after making a modification to ``lib/ansible/module_utils/docker_common.py``

I simply imported requests and printed a few diagnostic elements. The most pertinent result is what happens when you import requests:

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/var/folders/k5/y8mf3j355992gryff0n83j800000gn/T/ansible_Qruzwm/ansible_module_docker_container.py", line 655, in <module>
    from ansible.module_utils.docker_common import *
  File "/var/folders/k5/y8mf3j355992gryff0n83j800000gn/T/ansible_Qruzwm/ansible_modlib.zip/ansible/module_utils/docker_common.py", line 34, in <module>
ImportError: No module named requests
```

This is quite obviously not correct according to the environment that I *ran it from* because that does have requests:

```
# This is what docker_common.py actually does
$ python -c 'from requests.exceptions import SSLError'
$ echo $?
0
```

### Conclusion

So the original error message is technically not totally wrong, just the advice to ``pip install docker-py`` is wrong. So, a possible resolution to this would be to use ``ansible_python_interpreter=/path/to/valid/python`` in your hosts file for any host that needs to run docker..that's only a work around in my opinion and instead this should be fixed to include requests in the execution environment of ansible. I have not verified whether it ignores the whole environment or just requests for some strange reason.

**Comment by darkn3rd** [comment]
I did indeed install ansible with pip3 without knowing it.  Doing a `pip3 install --upgrade pip setuptools wheel` will unfortunately re-link pip to `pip3`, not leave it pointing to `pip2`.  Then when I tried to install with `pip`, it was actually doing it with `pip3`.  Frustrating, but now understood.  I removed brew install pythons and only upgraded the python2 environment.  This means my python3 will remain stale for now, but at least ansible works.

**Comment by nourspace** [comment]
To those using Anaconda, make sure to set this in your hosts file
```
localhost   ansible_connection=local    ansible_python_interpreter=/Users/USER/anaconda/bin/python
```
replace `USER` with your own user.

**Comment by waddles** [comment]
I hit this issue again today which gave me a chance to verify that the fix already mentioned [back in August](https://github.com/ansible/ansible-modules-core/issues/4246#issuecomment-238453608) is still the right solution. @pshahid this may help you:

Before making plays with the docker module, **make sure your remote host has docker-py installed**

Our playbook for building Ubuntu servers has these tasks:
```
- name: 'Install docker'
  apt:
    name: docker-engine
    update_cache: yes
    state: present

- name: 'Install pip'
  apt:
    name: python-pip
    state: present

- name: 'Install docker-py'
  pip:
    name: docker-py
    state: present
```
_I have omitted the preceding tasks to set up the dockerproject.org repo_

**Comment by francisdb** [comment]
For us the fix was setting up docker with this configuration:

```yml
  roles:
    - { role: angstwad.docker_ubuntu, become: yes, pip_version_docker_py: 1.9.0 }
```

**Comment by pshahid** [comment]
@waddles I understand that the remote host needs docker-py installed...seems like an odd thing for ansible to require. Is this a common requirement for other plugins as well?

**Comment by patran** [comment]
If you don't want to install docker-py on the target remote host, you would have to enable remote API (-H) for docker and use delegate_to the localhost. Ansible would then use the local docker-py to remotely connect with the target host. Before you enable the remote API, consider your security scenarios.

**Comment by dlazerka** [comment]
@patran Maybe it's possible to enable docker remote API on target, open it only for connections from 0.0.0.0, so Ansible will be able to ssh to target machine and talk to it?
